core/include/vecmem/containers/array.hpp

```diff
@@ -78,14 +85,14 @@ namespace vecmem {
       /// Can only be used if the user chose a non-default value for the size
       /// template parameter.
       ///
-      array( memory_resource& resource );
+      array( allocator_type alloc );
 
       /// Constructor with a size and a memory resource to use
       ///
       /// Can only be used if the user uses the default (invalid) value for the
       /// size template parameter.
       ///
-      array( memory_resource& resource, size_type size );
+      array( allocator_type alloc, size_type size );
```

Comment:
I'm not super happy about this formalism, but this seemed to be the most appropriate.

This class only needs the allocator object in its constructor. So it could declare it with a const reference as well. But since it doesn't make a copy of the allocator itself, and the `allocate(...)` function of the allocator is non-const, I decided that making the copy as part of the function signature was the way to go.

STL containers just hide this copy a bit more. (I was looking at the implementation of `std::vector` for quite a while to figure out how it passes the received allocator object around...)

Reply:
Yeah, I have spent more time than I would like to admit reading through the documentation to find out how exactly `std::pmr::vector` constructs itself using a memory resource. Either something is off with the documentation there, or there is some implicit construction of allocator happening. I must admit I don't really know.

Reply:
Note that now that `vecmem::array` uses an allocator instead of a memory resource directly, I even more strongly think that the order of these two constructor arguments should be reversed. So that it would provide the same interface as `vecmem::vector` when the size of the array is decided at runtime.

I was convinced that either you or @paulgessinger requested this argument order in #25. But I can't find that comment anymore. :confused: